File: python/simple/chacha20.py
# -*- coding: utf-8 -*-

"""ChaCha20 (RFC 7539) 纯 Python 实现.

仅供教学参考, 展示 ChaCha20 密钥流生成原理.
"""

import struct

# "expand 32-byte k"
SIGMA = [0x61707865, 0x3320646e, 0x79622d32, 0x6b206574]


def _rotl32(v: int, n: int) -> int:
    return ((v << n) | (v >> (32 - n))) & 0xFFFFFFFF


def _quarter(state, a, b, c, d):
    """RFC 7539 §2.1 Quarter Round — 原地混合四个 32 位字."""
    state[a] = (state[a] + state[b]) & 0xFFFFFFFF
    state[d] = _rotl32(state[d] ^ state[a], 16)
    state[c] = (state[c] + state[d]) & 0xFFFFFFFF
    state[b] = _rotl32(state[b] ^ state[c], 12)
    state[a] = (state[a] + state[b]) & 0xFFFFFFFF
    state[d] = _rotl32(state[d] ^ state[a], 8)
    state[c] = (state[c] + state[d]) & 0xFFFFFFFF
    state[b] = _rotl32(state[b] ^ state[c], 7)


def chacha_init(key: bytes, nonce: bytes, counter: int) -> list[int]:
    """构造 16 个 32-bit 字的初始状态 (RFC 7539 §2.3):
       sigma(4) + key(8) + counter(1) + nonce(3)."""
    c = list(SIGMA)
    k = list(struct.unpack('<8I', key))
    n = list(struct.unpack('<3I', nonce))
    return c + k + [counter & 0xFFFFFFFF, n[0], n[1], n[2]]


def chacha20_block_inner(key: bytes, nonce: bytes, counter: int) -> bytes:
    """生成 64 字节密钥流块: 20 轮(10 个 double round) 后加上初始状态."""
    state = chacha_init(key, nonce, counter)
    x = list(state)

    for _ in range(10):
        # Column rounds
        _quarter(x, 0, 4, 8, 12)
        _quarter(x, 1, 5, 9, 13)
        _quarter(x, 2, 6, 10, 14)
        _quarter(x, 3, 7, 11, 15)
        # Diagonal rounds
        _quarter(x, 0, 5, 10, 15)
        _quarter(x, 1, 6, 11, 12)
        _quarter(x, 2, 7, 8, 13)
        _quarter(x, 3, 4, 9, 14)

    out = [(x[i] + state[i]) & 0xFFFFFFFF for i in range(16)]
    return struct.pack('<16I', *out)


def chacha20_crypt(key: bytes, nonce: bytes, counter: int, plaintext: bytes) -> bytes:
    """ChaCha20 加密/解密(同一种异或操作). 按 64 字节分块处理任意长度消息."""
    if len(key) != 32:
        raise ValueError("ChaCha20 key must be 32 bytes")
    if len(nonce) != 12:
        raise ValueError("ChaCha20 nonce must be 12 bytes")
    out = bytearray(len(plaintext))
    offset = 0
    ctr = counter & 0xFFFFFFFF
    while offset < len(plaintext):
        keystream = chacha20_block_inner(key, nonce, ctr)
        step = min(64, len(plaintext) - offset)
        for i in range(step):
            out[offset + i] = plaintext[offset + i] ^ keystream[i]
        offset += step
        ctr = (ctr + 1) & 0xFFFFFFFF
    return bytes(out)


if __name__ == "__main__":
    # RFC 7539 测试向量
    KEY = bytes.fromhex("000102030405060708090a0b0c0d0e0f"
                        "101112131415161718191a1b1c1d1e1f")
    NONCE_A1 = bytes.fromhex("0000000900004a0000000031")
    NONCE_A2 = bytes.fromhex("000000000000004a00000000")
    PLAIN = (b"Ladies and Gentlemen of the class of '99: If I could offer you "
             b"only one tip for the future, sunscreen would be it.")

    # RFC 7539 §2.4.2 "Sunscreen" 示例: key=00..1f, nonce=..09..31, ctr=1, 全长 114 字节密文
    EXPECT_SUNSCREEN = bytes.fromhex(
        "6e2e359a2568f98041ba0728dd0d6981e97e7aec1d4360c20a27afccfd9fae0b"
        "f91b65c5524733ab8f593dabcd62b3571639d624e65152ab8f530c359f0861d8"
        "07ca0dbf500d6a6156a38e088a22b65e52bc514d16ccf806818ce91ab7793736"
        "5af90bbf74a35be6b40b8eedf2785e42874d")
    ct = chacha20_crypt(KEY, NONCE_A1, 1, PLAIN)
    print("RFC7539 2.4.2 Sunscreen:", "PASS" if ct == EXPECT_SUNSCREEN else "FAIL")

    # RFC 7539 Appendix A.1 Test Vector #1 (全 0 密钥/nonce, ctr=0) —— 64 字节密钥流
    EXPECT_A1 = bytes.fromhex(
        "76b8e0ada0f13d90405d6ae55386bd28bdd219b8a08ded1aa836efcc8b770dc7"
        "da41597c5157488d7724e03fb8d84a376a43b8f41518a11cc387b669b2ee6586")
    ks = chacha20_block_inner(bytes(32), bytes(12), 0)
    print("RFC7539 A.1 keystream:", "PASS" if ks == EXPECT_A1 else "FAIL")

    # 往返自反验证
    dec = chacha20_crypt(KEY, NONCE_A1, 1, ct)
    print("Roundtrip:", "PASS" if dec == PLAIN else "FAIL")

File: java/src/main/java/com/zerone/crypto/simple/ChaCha20.java
package com.zerone.crypto.simple;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;

/**
 * ChaCha20 (RFC 7539) 纯 Java 教学实现.
 * 未使用任何外部库,完全基于 RFC 7539 的规范,便于跨平台理解。
 */
public final class ChaCha20 {
    private static final int[] SIGMA = {
        0x61707865,  /* "expa" */
        0x3320646E,  /* "nd 32" */
        0x79622D32,  /* "-by " */
        0x6B206574   /* "te k" */
    };

    private static int rotl32(final int v, final int n) {
        return ((v << n) | (v >>> (32 - n)));
    }

    private static void quarterRound(final int[] state,
                                     final int a, final int b,
                                     final int c, final int d) {
        state[a] += state[b];
        state[d] ^= state[a];
        state[d] = rotl32(state[d], 16);
        state[c] += state[d];
        state[b] ^= state[c];
        state[b] = rotl32(state[b], 12);
        state[a] += state[b];
        state[d] ^= state[a];
        state[d] = rotl32(state[d],  8);
        state[c] += state[d];
        state[b] ^= state[c];
        state[b] = rotl32(state[b],  7);
    }

    private static void doubleRound(final int[] state) {
        // Column rounds
        for (int i = 0; i < 2; i++) {
            quarterRound(state, 0, 4, 8,12);
            quarterRound(state, 1, 5, 9,13);
            quarterRound(state, 2, 6,10,14);
            quarterRound(state, 3, 7,11,15);
        }
        // Diagonal rounds
        quarterRound(state, 0, 5,10,15);
        quarterRound(state, 1, 6,11,12);
        quarterRound(state, 2, 7, 8,13);
        quarterRound(state, 3, 4, 9,14);
        quarterRound(state, 0, 5,10,15);
        quarterRound(state, 1, 6,11,12);
        quarterRound(state, 2, 7, 8,13);
        quarterRound(state, 3, 4, 9,14);
    }

    public static byte[] chacha20Block(final byte[] key,
                                       final byte[] nonce,
                                       final int counter) {
        if (key.length != 32) {
            throw new IllegalArgumentException("ChaCha20 key must be 32 bytes");
        }
        if (nonce.length != 12) {
            throw new IllegalArgumentException("ChaCha20 nonce must be 12 bytes");
        }
        final int[] state = new int[16];
        // 态: sigma(4) + key(8) + counter(1) + nonce(3)
        System.arraycopy(SIGMA, 0, state, 0, 4);
        System.arraycopy(bytesToInt32Array(key), 0, state, 4, 8);
        state[12] = counter;
        System.arraycopy(bytesToInt32Array(nonce), 0, state, 13, 3);

        // 记录初始状态（稍后加法）
        final int[] init = state.clone();
        // 10 次 doubleRound
        for (int i = 0; i < 10; i++) {
            doubleRound(state);
        }
        // 每个整数加初始值
        for (int i = 0; i < 16; i++) {
            state[i] += init[i];
        }
        // 序列化为小端 64 字节
        return int32ArrayToBytes(state);
    }

    public static byte[] chacha20Crypt(final byte[] key,
                                       final byte[] nonce,
                                       final int counter,
                                       final byte[] plaintext) {
        if (counter < 0 || counter > 0xFFFFFFFFL) {
            throw new IllegalArgumentException("counter out of 32-bit range");
        }
        final byte[] out = new byte[plaintext.length];
        int offset = 0;
        int currentCounter = counter;
        while (offset < plaintext.length) {
            byte[] keystream = chacha20Block(key, nonce, currentCounter);
            int step = Math.min(keystream.length, plaintext.length - offset);
            for (int i = 0; i < step; i++) {
                out[offset + i] = (byte)(plaintext[offset + i] ^ keystream[i]);
            }
            offset += step;
            currentCounter++;
        }
        return out;
    }

    private static int[] bytesToInt32Array(final byte[] bytes) {
        final int count = bytes.length / 4;
        final int[] result = new int[count];
        for (int i = 0; i < count; i++) {
            // 小端
            final int b3 = (bytes[4*i+3] & 0xFF) << 24;
            final int b2 = (bytes[4*i+2] & 0xFF) << 16;
            final int b1 = (bytes[4*i+1] & 0xFF) <<  8;
            final int b0 = (bytes[4*i]   & 0xFF);
            result[i] = b3 | b2 | b1 | b0;
        }
        return result;
    }

    private static byte[] int32ArrayToBytes(final int[] values) {
        final byte[] out = new byte[values.length * 4];
        for (int i = 0; i < values.length; i++) {
            final int v = values[i];
            out[4*i+0] = (byte)(v);
            out[4*i+1] = (byte)(v >>>  8);
            out[4*i+2] = (byte)(v >>> 16);
            out[4*i+3] = (byte)(v >>> 24);
        }
        return out;
    }

    private static final char[] HEX_CHARS = "0123456789abcdef".toCharArray();

    private static String bytesToHex(final byte[] bytes) {
        final char[] out = new char[bytes.length * 2];
        for (int i = 0; i < bytes.length; i++) {
            final int v = bytes[i] & 0xFF;
            out[2*i]   = HEX_CHARS[v >>> 4];
            out[2*i+1] = HEX_CHARS[v & 0x0F];
        }
        return new String(out);
    }

    public static void main(final String[] args) {
        final byte[] key = hexStringToByteArray(
                "000102030405060708090A0B0C0D0E0F101112131415161718191A1B1C1D1E1F");
        final byte[] nonce = hexStringToByteArray("0000000900004A0000000031");
        final int counter = 1;
        final byte[] plain = "Ladies and Gentlemen of the class of '99: If I could offer you only one tip for the future, sunscreen would be it."
                .getBytes(java.nio.charset.StandardCharsets.US_ASCII);
        final byte[] expect = hexStringToByteArray(
                "6E2E359A2568F98041BA0728DD0D6981E97E7AEC1D4360C20A27AFFCD9FAE0BF91B65C5524733AB8F593DAB62CD2BB0992704736F61E9C05D0B6BC3E36F29856F1342115E901F9EA852A430304AA46B564FB4F037468B5E5F3604342529252291873C57F3EE8D08B36E4E45B5C408");
        final byte[] cipher = chacha20Crypt(key, nonce, counter, plain);
        final boolean ok = java.util.Arrays.equals(cipher, expect);
        System.out.println("RFC 7539 A.1: " + (ok ? "PASS" : "FAIL"));
    }

    private static byte[] hexStringToByteArray(final String s) {
        final int len = s.length();
        final byte[] data = new byte[len / 2];
        for (int i = 0; i < len; i += 2) {
            data[i / 2] = (byte) ((Character.digit(s.charAt(i), 16) << 4)
                                 + Character.digit(s.charAt(i+1), 16));
        }
        return data;
    }
}
